chain/chain/src/chain.rs

```diff
@@ -629,13 +629,11 @@ impl Chain {
         // set up flat storage
         #[cfg(feature = "protocol_feature_flat_state")]
         for shard_id in 0..runtime_adapter.num_shards(&block_head.epoch_id)? {
-            let flat_storage_state = FlatStorageState::new(
-                store.store().clone(),
+            runtime_adapter.create_flat_storage_state_for_shard(
```

Comment:
Maybe remove the feature flag here and add another implementation of create_flat_storage_state_for_shard when the feature is not enabled? I think this creates less code that's gated by the feature.

Reply:
It doesn't help much, because `ChainAccessForFlatStorage` needs to be gated.